include/cuco/priority_queue.cuh

```diff
@@ -0,0 +1,239 @@
+    {
+      int intersection_bytes = 2 * (block_size + 1) * sizeof(int);
+      int node_bytes         = node_size_ * sizeof(T);
+      return intersection_bytes + 2 * node_bytes;
+    }
+
+    __host__ __device__ device_mutable_view(size_t node_size,
+                                            T* d_heap,
+                                            int* d_size,
+                                            size_t* d_p_buffer_size,
+                                            int* d_locks,
+                                            int lowest_level_start,
+                                            int node_capacity,
+                                            Compare const& compare)
+      : node_size_(node_size),
+        d_heap_(d_heap),
+        d_size_(d_size),
+        d_p_buffer_size_(d_p_buffer_size),
+        d_locks_(d_locks),
+        lowest_level_start_(lowest_level_start),
+        node_capacity_(node_capacity),
+        compare_(compare)
+    {
+    }
+
+   private:
+    size_t node_size_;
+    int lowest_level_start_;
+    int node_capacity_;
+
+    T* d_heap_;
+    int* d_size_;
+    size_t* d_p_buffer_size_;
+    int* d_locks_;
+    Compare compare_;
+  };
+
+  /*
+   * @brief Returns a trivially-copyable class that can be used to perform
+   *        insertion and deletion of single nodes in device code with
+   *        cooperative groups
+   *
+   * @return A device view
+   */
+  device_mutable_view get_mutable_device_view()
+  {
+    return device_mutable_view(node_size_,
+                               d_heap_,
+                               d_size_,
+                               d_p_buffer_size_,
+                               d_locks_,
+                               lowest_level_start_,
+                               node_capacity_,
+                               compare_);
+  }
+
+ private:
+  size_t node_size_;        ///< Size of the heap's nodes
+  int lowest_level_start_;  ///< Index in `d_heap_` of the first node in the
+                            ///  heap's lowest level
+  int node_capacity_;       ///< Capacity of the heap in nodes
+
+  T* d_heap_;                ///< Pointer to an array of nodes, the 0th node
+                             ///  being the heap's partial buffer, and nodes
+                             ///  1..(node_capacity_) being the heap, where the
+                             ///  1st node is the root
+  int* d_size_;              ///< Number of nodes currently in the heap
+  size_t* d_p_buffer_size_;  ///< Number of elements currently in the partial
+                             ///  buffer
+  int* d_locks_;             ///< Array of locks where `d_locks_[i]` is the
+                             ///  lock for the node starting at
+                             ///  1d_heap_[node_size * i]`
+
+  Allocator allocator_;
+  int_allocator_type int_allocator_;
+  t_allocator_type t_allocator_;
+  size_t_allocator_type size_t_allocator_;
+
+  Compare compare_{};
```

Comment:
```suggestion
```
`Compare` should be a template parameter for public APIs like `push`/`pop` instead of an attribute in `priority_queue` or its view. e.g.
```cpp
template <typename InputIt, typename Equal = thrust::less<T>>
void push(InputIt first, InputIt last, Equal equal = Equal{}, cudaStream_t stream = 0); 
```

Reply:
I don't think this would be legitimate, as the heap structure where the elements are stored keeps them ordered according to a certain comparator. In what case might a user want to supply different comparators for different operations?